maven/spec/dependabot/maven/version_spec.rb

```diff
@@ -115,295 +145,350 @@
+          expect(other_version <=> version).to eq(-result)
         end
       end
+    end
 
-      context "when greater" do
-        let(:other_version) { described_class.new("1.1.0") }
-
-        it { is_expected.to eq(-1) }
-      end
-
-      context "when the version is a post-release" do
-        let(:other_version) { described_class.new("1.0.0u1") }
+    context "with semantic versions that have a qualifier" do
+      let(:versions) do
+        [
+          ["1.2.3", "1.2.3-a0", 1], # alpha has lower precedence
+          ["1.2.3", "1.2.3-a", -1] # 'a' without a following int is not alpha
 
-        it { is_expected.to eq(-1) }
+        ]
       end
 
-      context "when the version is a pre-release" do
-        let(:other_version) { described_class.new("1.0.0a1") }
-
-        it { is_expected.to eq(1) }
+      it "returns the correct result" do
+        versions.each do |input|
+          version1, version2, result = input
+          version = described_class.new(version1)
+          other_version = described_class.new(version2)
+          expect(version <=> other_version).to eq(result)
+          expect(other_version <=> version).to eq(-result)
+        end
       end
+    end
 
-      context "when the version is non-numeric" do
-        let(:version) { described_class.new("Finchley") }
-        let(:other_version) { described_class.new("Edgware") }
-
-        it { is_expected.to eq(1) }
+    context "with versions that have trailing nulls" do
+      let(:versions) do
+        [
+          ["1alpha-0", "1alpha", 0],
+          ["1alpha-0", "1alpha0", 0],
+          ["1alpha-0", "1alpha.0", 0],
+          ["1alpha-0", "1alpha.z", -1],
+
+          ["1beta-0", "1beta", 0],
+          ["1beta-0", "1beta0", 0],
+          ["1beta-0", "1beta.0", 0],
+          ["1beta-0", "1beta.z", -1],
+
+          ["1rc-0", "1rc0", 0],
+          ["1rc-0", "1rc", 0],
+          ["1rc-0", "1rc.0", 0],
+          ["1rc-0", "1rc.z", -1],
+
+          ["1sp-0", "1sp0", 0],
+          ["1sp-0", "1sp", 0],
+          ["1sp-0", "1sp.0", 0],
+          ["1sp-0", "1sp.z", -1],
+
+          ["1.ga", "1-ga", 0],
+          ["1.0", "1.ga", 0],
+          ["1.0.FINAL", "1", 0],
+          ["1.0", "1.release", 0],
+
+          ["1.2.0", "1.2", 0],
+          ["1.2.3", "1.2.3-0", 0],
+          ["1.2.3-0", "1.2.3-0", 0],
+          ["1.2.3-0", "1.2.3-a0", 1],
+          ["1.2.3-0", "1.2.3-a", -1],
+          ["1.2.3-0", "1.2.3-1", -1],
+          ["1.2.3-0", "1.2.3-0-1", -1],
+
+          ["1snapshot-0", "1snapshot0", 0],
+          ["1snapshot-0", "1snapshot", 0],
+          ["1snapshot-0", "1snapshot.0", 0],
+          ["1snapshot-0", "1snapshot.z", -1],
+
+          ["1milestone-0", "1milestone", 0],
+          ["1milestone-0", "1milestone0", 0],
+          ["1milestone-0", "1milestone.0", 0],
+          ["1milestone-0", "1milestone.z", -1]
+        ]
       end
 
-      describe "with a + separated alphanumeric build identifier" do
```

Comment:
Moved to context 'when comparing numerically'. By the way, unlike in semver 2 where '+' is followed by a build number, in maven it's considered just a character.